### PR TITLE
[POC] More precise error catching in `doPoll`

### PR DESCRIPTION
**Pull request by guizmaii**

### CODE REVIEW

File: zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala [hunk 1]
```diff
@@ -331,7 +332,7 @@ private[consumer] final class Runloop(
       try
         c.poll(pollTimeout)
       catch {
-        case _: IllegalStateException =>
+        case _: IllegalStateException if isUnsubscribed =>
```
Comment (erikvanoosten) [inline zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala:335]: I would guess that you would have the same race condition here. So is `isUnsubscribed` reliable?

Comment (guizmaii) [reply zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala:335]: Hum 🤔

Comment (guizmaii) [reply zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala:335]: Should we even try to `c.poll` if we're unsubscribed? 🤔

File: zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala [hunk 2]
```diff
@@ -36,7 +37,7 @@ private[consumer] final class Runloop(
   shutdownRef: Ref[Boolean],
   offsetRetrieval: OffsetRetrieval,
   userRebalanceListener: RebalanceListener,
-  subscribedRef: Ref[Boolean],
+  subscribedRef: AtomicBoolean,
```
Comment (svroonland) [inline zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala:40]: I would prefer to stick to ZIO Refs here and see if we can find a good solution using that. Somehow feels safer and more zio-idiomatic.

File: zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala (hunk 1 above)
Comment (svroonland) [reply zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala:335]: I just noticed that the call to `handlePoll` in `handleShutdown` is not guarded by the `isSubscribed` check, perhaps that was the cause of the race condition..

**Comment by svroonland** [comment]
Thanks for helping along here. I've been thinking, if we want to avoid changing the subscription in the middle of `handlePoll`, we should introduce a Command so we can handle it sequentially. I'd like to investigate that direction

**Comment by guizmaii** [comment]
> we should introduce a Command so we can handle it sequentially. I'd like to investigate that direction

Interesting idea. Please do :)

Thanks for your amazing work! ❤️

**Comment by svroonland** [comment]
@guizmaii @erikvanoosten I implemented the Command idea here: https://github.com/zio/zio-kafka/pull/517/commits/85378e72c6861c754fd613b7de4ccab64dc500ae

**Pull request closed without merging**